### PR TITLE
RUM-3634: Apply privacy settings to TextCompositionGroupMapper for compose

### PR DESCRIPTION
**Pull request by ambushwork**
### What does this PR do?

Apply the privacy settings to `TextCompositionGroupMapper` for compose, the text masking strategy keeps same as `TextViewMapper` in Android View

### Motivation

* RUM-3634

### Screenrecord

| USER_INPUT | MASK | 
|---|---|
|https://mobile-integration.datadoghq.com/rum/replay/sessions/e55f5ad0-145d-4c2c-bc5b-1c4bbcd35ba9?applicationId=38030dde-f9f9-4e52-9443-b9804a030080&seed=70a2b95d-ba66-4083-ad43-81134787c57b&ts=1720118965911 | https://mobile-integration.datadoghq.com/rum/replay/sessions/5b0a74de-9928-4fb6-a273-e5d2a4f7ded3?applicationId=38030dde-f9f9-4e52-9443-b9804a030080&seed=49b1537e-3107-4da9-b7a6-49c24ed46145&ts=1720119348991|

### Additional Notes

* `StringObfuscator` is changed from `internal` class to `@InternalApi`, so the class can be shared between module.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapper.kt [hunk 1]
```diff
@@ -79,6 +82,22 @@ internal class TextCompositionGroupMapper(
...
+            SessionReplayPrivacy.ALLOW -> originalText
+            SessionReplayPrivacy.MASK_USER_INPUT ->
+                if (uiContext.hasOptionSelectorParent) FIXED_INPUT_MASK else originalText
+
+            SessionReplayPrivacy.MASK ->
+                if (uiContext.hasOptionSelectorParent) {
+                    FIXED_INPUT_MASK
+                } else {
+                    StringObfuscator.getStringObfuscator().obfuscate(originalText)
+                }
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapper.kt:97]: I know that `Spinner` or likewise components don't exist in Compose, but if I understand correctly, the first way you propose, we still need to keep this information in `UiContext` (maybe just need to change the naming). The thing I need to change is to remove the comment of "TODO" for providing `OptionSelectorDetector` for it, because we are not able to get this information with the input of `ViewGroup`, we need to have another way just like you proposed.

About the second proposition, we are just discussing "Fine Grained Masking" for Q3, it might some crossover with this idea. and that might be a good one.

File: features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/ComposeExtensionSupport.kt [hunk 2]
```diff
@@ -45,6 +45,7 @@ class ComposeExtensionSupport : ExtensionSupport {
     }
 
     override fun getOptionSelectorDetectors(): List<OptionSelectorDetector> {
+        // TODO RUM-5224: Implement compose OptionSelectorDetector to detect option text as user input
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/ComposeExtensionSupport.kt]: removed

File: features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/data/UiContext.kt [hunk 3]
```diff
@@ -7,10 +7,13 @@
...
 
 import androidx.compose.ui.unit.Density
+import com.datadog.android.sessionreplay.SessionReplayPrivacy
 
 internal data class UiContext(
     val parentContentColor: String?,
-    val density: Float
+    val density: Float,
+    val privacy: SessionReplayPrivacy,
+    val hasOptionSelectorParent: Boolean = false
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/data/UiContext.kt]: the field has been renamed to `isInUserInputLayout`

File: features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/ComposeWireframeMapper.kt [hunk 4]
```diff
@@ -60,12 +61,18 @@ internal class ComposeWireframeMapper(
         internalLogger: InternalLogger
     ): List<MobileSegment.Wireframe> {
         val density = mappingContext.systemInformation.screenDensity.let { if (it == 0.0f) 1.0f else it }
-
+        val privacy = mappingContext.privacy
+        val hasOptionSelectorParent = mappingContext.hasOptionSelectorParent
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/ComposeWireframeMapper.kt]: removed,

**Review by xgouchet — APPROVED** [review]

**Comment by ambushwork** [comment]
@xgouchet  I just rebased the branch due to some merge conflicts

File: features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    private lateinit var mockCompositionGroup: CompositionGroup
+
+    @BeforeEach
+    fun `set up`(forge: Forge) {
+        textCompositionGroupMapper = TextCompositionGroupMapper(colorStringFormatter = DefaultColorStringFormatter)
+        mockCompositionGroup = mockGroupWithCoordinates(forge)
+    }
+
+    @RepeatedTest(8)
+    fun `M return to correct wireframe W map`() {
```
Comment (jonathanmos) [inline features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt:50]:
```suggestion
    fun `M return the correct wireframe W map`() {
```

File: sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt [hunk 6]
```diff
@@ -151,7 +151,7 @@ class SampleApplication : Application() {
                     useCustomEndpoint(BuildConfig.DD_OVERRIDE_SESSION_REPLAY_URL)
                 }
             }
-            .setPrivacy(SessionReplayPrivacy.MASK_USER_INPUT)
+            .setPrivacy(SessionReplayPrivacy.MASK)
```
Comment (jonathanmos) [inline sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt:154]: minor: what is the reason for this change?

**Review by jonathanmos — APPROVED** [review]

File: sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt (hunk 6 above)
Comment (ambushwork) [reply sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt]: for the test purpose, it shouldn't be in the commit, reverted now

File: features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            composeContext = fakeComposeContext,
+            uiContext = fakeUiContext
+        )
+        val expectedText = when (fakeUiContext.privacy) {
+            SessionReplayPrivacy.ALLOW -> String()
+            SessionReplayPrivacy.MASK,
+            SessionReplayPrivacy.MASK_USER_INPUT -> if (fakeUiContext.isInUserInputLayout) {
+                FIXED_INPUT_MASK
+            } else {
+                String()
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt]: personal habit of giving a empty string : ) now fixed

File: features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt (hunk 5 above)
Comment (ambushwork) [reply features/dd-sdk-android-session-replay-compose/src/test/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/TextCompositionGroupMapperTest.kt]: updated

**Review by xgouchet — APPROVED** [review]
